src/index.ts

```diff
@@ -15,62 +25,115 @@ const buildRestClient = (apiClient: InfisicalApi, requestOptions?: RawAxiosReque
 	};
 };
 
-// We need to do bind(this) because the authenticate method is a private method, and usually you can't call private methods from outside the class.
-type InfisicalSDKOptions = {
-	siteUrl?: string;
-};
-
 class InfisicalSDK {
 	#apiInstance: InfisicalApi;
 
+	sdkOptions: InfisicalSDKOptions;
+
 	#requestOptions: RawAxiosRequestConfig | undefined;
 	#secretsClient: SecretsClient;
 	#dynamicSecretsClient: DynamicSecretsClient;
 	#authClient: AuthClient;
-	#basePath: string;
+	#tokenDetails?: TTokenDetails;
+	#authCredentials?: TAuthCredentials;
 
 	constructor(options?: InfisicalSDKOptions) {
-		this.#basePath = options?.siteUrl || "https://app.infisical.com";
+		this.sdkOptions = {
+			autoTokenRefresh: options?.autoTokenRefresh ?? true,
+			siteUrl: options?.siteUrl || "https://app.infisical.com"
+		};
 
 		this.#apiInstance = new InfisicalApi(
 			new Configuration({
-				basePath: this.#basePath
+				basePath: this.sdkOptions.siteUrl
 			})
 		);
 
-		this.#authClient = new AuthClient(this.authenticate.bind(this), this.#apiInstance, this.#basePath);
+		// We need to do bind(this) because the authenticate method is a private method, and usually you can't call private methods from outside the class.
+		this.#authClient = new AuthClient(this.#authenticator, this.#apiInstance, this.#tokenDetails?.accessToken);
 		this.#dynamicSecretsClient = new DynamicSecretsClient(this.#apiInstance, this.#requestOptions);
 		this.#secretsClient = new SecretsClient(this.#apiInstance, this.#requestOptions);
 		this.rest = () => buildRestClient(this.#apiInstance, this.#requestOptions);
+
+		if (this.sdkOptions.autoTokenRefresh) {
+			this.handleTokenLifecycle();
+		}
 	}
 
-	private authenticate(accessToken: string) {
-		this.#apiInstance = new InfisicalApi(
-			new Configuration({
-				basePath: this.#basePath,
-				accessToken
-			})
-		);
+	private handleTokenLifecycle = async () => {
+		const authMethodMap: Record<AuthMethod, () => Promise<InfisicalSDK>> = {
+			[AuthMethod.UniversalAuth]: async () =>
+				await this.#authClient.universalAuth.login(this.#authCredentials?.credentials as UniversalAuthCredentials),
+			[AuthMethod.AWSIam]: async () => await this.#authClient.awsIamAuth.login(this.#authCredentials?.credentials as AWSIamCredentials),
+			[AuthMethod.AccessToken]: async () =>
+				await this.#authClient.accessToken((this.#authCredentials?.credentials as AccessTokenCredentials).accessToken)
+		};
+
+		while (true) {
```

Comment:
I'm worried about how memory safe this is. i know in go we have go routine but unsure how this compares to a every running while loop

Reply:
1. Excessive CPU usage is limited by the sleep function
2. The loop will never break due to the try-catch statement
3. while(true) is necessary in this case, because intended behavior is to have this run in the background. Javascript isn't multi-threaded, and we don't want to spawn a sub-process for handling token lifecycle